src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs

```diff
@@ -345,7 +346,28 @@ private HttpMessageHandler SetupHandlerChain()
                 return Task.FromException<HttpResponseMessage>(error);
             }
 
-            return handler.SendAsync(request, cancellationToken);
+            return HttpTelemetry.IsEnabled && request.RequestUri != null ?
+                WithLogging(handler, request, cancellationToken) :
+                handler.SendAsync(request, cancellationToken);
+
+            static async Task<HttpResponseMessage> WithLogging(HttpMessageHandler handler, HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                Debug.Assert(request.RequestUri != null);
+                HttpTelemetry.Log.RequestStart(request.RequestUri.IdnHost, request.RequestUri.Port);
+                try
+                {
+                    return await handler.SendAsync(request, cancellationToken).ConfigureAwait(false);
+                }
+                catch
+                {
+                    HttpTelemetry.Log.RequestAbort(request.RequestUri.IdnHost, request.RequestUri.Port);
+                    throw;
+                }
+                finally
+                {
+                    HttpTelemetry.Log.RequestStop(request.RequestUri.IdnHost, request.RequestUri.Port);
```

Comment:
I brought up [in another issue](https://github.com/dotnet/runtime/issues/37141) that Http telemetry appears to count receiving a message's headers as the request being done, not when the request is finished.

Is RequestStop the right name for this?

Is current requests the right counter when a request can still be in-progress?

Reply:
The 'Start' and 'Stop' suffixes are required in order for the EventSource activity threading to work automatically.  See comments in source code in the PR.

As discussed in #37141, HttpClient and the handler architecture define a request as 'done' when the request headers arrive.  Download of the response body is outside of the handlers themselves.

Reply:
> As discussed in #37141, HttpClient and the handler architecture define a request as 'done' when the request headers arrive. Download of the response body is outside of the handlers themselves.

We could make it work if we wanted to, albeit at an expense.  We could wrap the returned stream in a custom one that passed through the calls to the wrapped stream, and then signaled Stop when EOF is hit.

Reply:
>We could make it work if we wanted to

Do we want to? This is an interesting idea. We would have to make it work across HTTP versions (1.x, 2, 3).

Perhaps I can file a separate issue for this. For now, we can do it as I've done in the PR.  But use the separate issue to further discuss this topic and perhaps revise accordingly?

Reply:
Separate issue sounds fine. thanks.

Reply:
The connection pool could probably better track active requests end-to-end by tracking http/1.x connections and h2 streams in use. Tracking stop and current-requests at the SocketHttpHandler level gives confusing results.

Reply:
I added it to SocketsHttpHandler because it maps better to how the public SendAsync API works and it avoids having to add it in multiple places for HTTP 1.x, 2, 3 versions.

>Tracking stop and current-requests at the SocketHttpHandler level gives confusing results.

I don't understand why you consider it confusing.

Reply:
> HttpClient and the handler architecture define a request as 'done' when the request headers arrive. Download of the response body is outside of the handlers themselves.

HttpClient buffers most responses by default, those aren't considered done until fully buffered.

The naming/namespace of RequestStart/Stop and current-requests does not specify that these are scoped to the HttpClient or SocketHttpHandler. They sound like end-to-end lifetime events. Maybe the EventSource name for these should be changed to System.Net.Http.SocketHttpHandler.

Reply:
I agree with @Tratcher. We have following scenarios:

- SendAsync unbuffered
- SendAsync buffered
- Convenience methods (buffered)

It will be confusing for the latter two if the timings don't measure the entire operation. We should consider separate "headers received" and "request 100% finished" timings.

Reply:
>We should consider separate "headers received" and "request 100% finished" timings.

I agree we can add separate events for things like "headers received" and "response body fully read". Those can be sprinkled into SocketsHttpHandler (we already have diagnostic events at similar points).

But we still need to provide a definitive 'Stop' event for activity threading. So, are we thinking that we want 'RequestStop' to always mean response headers read and response body download complete?

Reply:
> But we still need to provide a definitive 'Stop' event for activity threading. So, are we thinking that we want 'RequestStop' to always mean response headers read and response body download complete?

From a logging / counters perspective, I'm not sure that having received response headers is particularly relevant.  Seems by default we should be tracking the whole request/response; if there's an advanced scenario that wants to be notified of just headers having been received, that could be done, too, in some fashion.

Reply:
Even if we do not buffer response, the stream we create does know when the response is done, right? (and we try to drain if we hand out stream and that is not read to the end)

Reply:
Based on this discussion and the consensus of everyone involved, I will change the PR so that 'RequestStop' includes reading the response body until EOF (or error).

I will also re-activate #37141. But that issue was about the DiagnosticSource telemetry in DiagnosticHandler.